pkg/agent/agent.go

```diff
@@ -0,0 +1,610 @@
+	// If node-specific configuration was added or updated switch to it.
+	// If it was removed switch to group-specific one if we have it.
+	//
+
+	if cfg == nil {
+		cfg = a.groupCfg
+	}
+	if cfg == nil {
+		log.Warnf("node (%s) has no effective configuration", a.nodeName)
+		return
+	}
+
+	err := a.notifyFn(cfg)
+	if err != nil {
+		log.Errorf("failed to apply configuration: %v", err)
+	}
+
+	a.patchConfigStatus(a.currentCfg, cfg, err)
+	a.currentCfg = cfg
+}
+
+func (a *Agent) updateGroupConfig(obj runtime.Object) {
+	var cfg cfgapi.AnyConfig
+
+	if obj != nil {
+		cfg = obj.(cfgapi.AnyConfig)
+	}
+
+	if sameConfigVersion(cfg, a.groupCfg) {
+		log.Debug("ignoring duplicate group-specific config update")
+		return
+	}
+
+	if cfg == nil {
+		log.Info("group-specific config deleted")
+	} else {
+		log.Info("group-specific config updated")
+	}
+
+	//
+	// update group-specific configuration
+	//
+	// If we don't have node-specific configuration but we have group-
+	// specific one, switch to it and notify about it.
+	//
+
+	a.groupCfg = cfg
+
+	if a.nodeCfg != nil {
+		return
+	}
+
+	if cfg == nil {
+		log.Warnf("node (%s) has no effective configuration", a.nodeName)
+		return
+	}
+
+	err := a.notifyFn(cfg)
+	if err != nil {
+		log.Errorf("failed to apply configuration: %v", err)
+	}
+
+	a.patchConfigStatus(a.currentCfg, cfg, err)
+	a.currentCfg = cfg
+}
+
+func (a *Agent) patchConfigStatus(prev, curr cfgapi.AnyConfig, err error) {
+	if a.cfgIf == nil {
+		return
+	}
+
+	var (
+		ctx      = context.TODO()
+		ns       = a.namespace
+		node     = a.nodeName
+		prevName string
+		currName string
+	)
+
+	if prev != nil {
+		prevName = prev.GetObjectMeta().Name
+	}
+	if curr != nil {
+		currName = curr.GetObjectMeta().Name
+	}
+
+	if prev != nil && prevName != currName {
+		pErr := a.cfgIf.PatchStatus(ctx, ns, prevName, node, nil)
+		if pErr != nil {
+			log.Errorf("failed to patch status of previous config %s/%s: %v", ns, prevName, pErr)
+		}
+	}
+
+	if curr != nil {
+		status := cfgapi.NewNodeStatus(err, curr.GetObjectMeta().Generation)
+		pErr := a.cfgIf.PatchStatus(ctx, ns, currName, node, status)
+		if pErr != nil {
+			log.Errorf("failed to patch status of current config %s/%s: %v", ns, currName, pErr)
+		}
+	}
+}
+
+func sameConfigVersion(cfg1, cfg2 cfgapi.AnyConfig) bool {
+	switch {
+	case cfg1 == nil && cfg2 == nil:
+		return true
+	case cfg1 == nil && cfg2 != nil:
+		return false
+	case cfg1 != nil && cfg2 == nil:
+		return false
+	}
+
+	o1 := cfg1.GetObjectMeta()
+	o2 := cfg2.GetObjectMeta()
+
+	switch {
+	case o1.UID != o2.UID:
+		return false
+	case o1.Generation != o2.Generation:
+		return false
+	case o1.Generation == 0: // config from a file
+		return false
+	}
+
+	return true
+}
+
+var (
+	defaultNamespace  string
+	defaultGroupLabel string
+	defaultKubeConfig string
+	defaultConfigFile string
+)
+
+func init() {
+	flag.StringVar(&defaultNamespace, "config-namespace", "kube-system",
+		"namespace for configuration CustomResources")
```

Comment:
I think our current Helm charts in main/HEAD and the patched version of those Helm charts in this PR does not provide an explicit way to opt out from installing ConfigMaps or custom resources for configuration.

The charts install a default ConfigMap/custom resource, so the user has two ways of overriding the configuration. One is the traditional Helm way of providing custom values for Helm, since the charts are written so that the configuration is templated. The other way is to install a node- or group specific configuration that overrides the defaults.

In all cases, the agent only monitors the indicated configuration namespace (which defaults to kube-system) and nothing else. Also, just like there is no way to dynamically move a daemonset from one namespace to another, we neither need (IMO) nor provide a way to dynamically change the namespace for our configuration custom resources (which the agent monitors). The Helm charts are deliberately written so that the same namespace is used for daemonsets and the configuration resources.